go/admin/server/server.go

```diff
@@ -115,3 +119,171 @@ func (s *MultiAdminServer) GetDatabase(ctx context.Context, req *multiadminpb.Ge
+
+// GetDatabaseNames retrieves all database names in the cluster
+func (s *MultiAdminServer) GetDatabaseNames(ctx context.Context, req *multiadminpb.GetDatabaseNamesRequest) (*multiadminpb.GetDatabaseNamesResponse, error) {
+	s.logger.Debug("GetDatabaseNames request received")
+
+	names, err := s.ts.GetDatabaseNames(ctx)
+	if err != nil {
+		s.logger.Error("Failed to get database names from topology", "error", err)
+		return nil, status.Errorf(codes.Internal, "failed to retrieve database names: %v", err)
+	}
+
+	response := &multiadminpb.GetDatabaseNamesResponse{
+		Names: names,
+	}
+
+	s.logger.Debug("GetDatabaseNames request completed successfully", "count", len(names))
+	return response, nil
+}
+
+// GetGateways retrieves gateways filtered by cells
+func (s *MultiAdminServer) GetGateways(ctx context.Context, req *multiadminpb.GetGatewaysRequest) (*multiadminpb.GetGatewaysResponse, error) {
+	s.logger.Debug("GetGateways request received", "cells", req.Cells)
+
+	// Determine which cells to query
+	cellsToQuery := req.Cells
+	if len(cellsToQuery) == 0 {
+		// If no cells specified, get all cells
+		allCells, err := s.ts.GetCellNames(ctx)
+		if err != nil {
+			s.logger.Error("Failed to get all cell names", "error", err)
+			return nil, status.Errorf(codes.Internal, "failed to retrieve cell names: %v", err)
+		}
+		cellsToQuery = allCells
+	}
+
+	var allGateways []*clustermetadatapb.MultiGateway
+
+	// Query each cell for gateways
+	for _, cellName := range cellsToQuery {
+		gatewayInfos, err := s.ts.GetMultiGatewaysByCell(ctx, cellName)
+		if err != nil {
+			s.logger.Error("Failed to get gateways for cell", "cell", cellName, "error", err)
+			// Continue with other cells instead of failing completely
+			continue
+		}
+
+		// Convert to protobuf
+		for _, info := range gatewayInfos {
+			gateway := info.MultiGateway
+			allGateways = append(allGateways, gateway)
+		}
+	}
+
+	response := &multiadminpb.GetGatewaysResponse{
+		Gateways: allGateways,
+	}
+
+	s.logger.Debug("GetGateways request completed successfully", "count", len(allGateways))
+	return response, nil
+}
+
+// GetPoolers retrieves poolers filtered by cells and/or database
+func (s *MultiAdminServer) GetPoolers(ctx context.Context, req *multiadminpb.GetPoolersRequest) (*multiadminpb.GetPoolersResponse, error) {
+	s.logger.Debug("GetPoolers request received", "cells", req.Cells, "database", req.Database)
+
+	// Determine which cells to query
+	cellsToQuery := req.Cells
+	if len(cellsToQuery) == 0 {
+		// If no cells specified, get all cells
+		allCells, err := s.ts.GetCellNames(ctx)
+		if err != nil {
+			s.logger.Error("Failed to get all cell names", "error", err)
+			return nil, status.Errorf(codes.Internal, "failed to retrieve cell names: %v", err)
+		}
+		cellsToQuery = allCells
+	}
+
+	var allPoolers []*clustermetadatapb.MultiPooler
+
+	// Query each cell for poolers
+	for _, cellName := range cellsToQuery {
+		poolerInfos, err := s.ts.GetMultiPoolersByCell(ctx, cellName, nil)
```

Comment:
This method already accepts options that will filter by Database. So we can push down the filter there.